pkg/engine/decisions.go

```diff
@@ -0,0 +1,226 @@
+package engine
+
+import (
+	"fmt"
+
+	"github.com/klothoplatform/klotho/pkg/construct"
+	"github.com/klothoplatform/klotho/pkg/engine/constraints"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base"
+)
+
+type (
+	// Decision is a struct that represents a decision made by the engine
+	Decision struct {
+		Level  Level
+		Result *DecisionResult
+		Action Action
+		Cause  *Cause
+	}
+	Action string
+	Cause  struct {
+		EdgeExpansion         *graph.Edge[construct.Resource]
+		EdgeConfiguration     *graph.Edge[construct.Resource]
+		OperationalResource   construct.Resource
+		ResourceConfiguration construct.Resource
+		ConstructExpansion    construct.BaseConstruct
+		Constraint            constraints.Constraint
+	}
+	DecisionResult struct {
+		Resource construct.Resource
+		Edge     *graph.Edge[construct.Resource]
+		Config   *knowledgebase.Configuration
+	}
+	Level string
+)
+
+const (
+	// LevelError is the level of a decision when an error occurs
+	LevelError Level = "error"
+	// LevelInfo is the level of a decision when an informational decision is made
+	LevelInfo Level = "info"
+	// LevelDebug is the level of a decision when a debug decision is made
+	LevelDebug Level = "debug"
+	// LevelWarn is the level of a decision when a warning decision is made
+	LevelWarn Level = "warn"
+
+	ActionConfigure  Action = "configure"
+	ActionDelete     Action = "delete"
+	ActionCreate     Action = "create"
+	ActionConnect    Action = "connect"
+	ActionDisconnect Action = "disconnect"
+)
+
+func (context *SolveContext) recordDecision(decision Decision) {
+	// Right now this is a wrapper around append, but we may want to do more in the future
+	//
+	// For example, we may want to check if there are duplicate decisions in the stack/history of decisions. If so we can likely determine that we are looping or going to fail and can shortcuircit the operations which are in a loop.
+	context.Decisions = append(context.Decisions, decision)
+}
+
+// Here we want to validate that each decision is crafted correctly and that there is a valid cause, action and result pairing
+func (d Decision) Validate() error {
```

Comment:
Personally, I don't see the use in validating structs that we're creating internally (ie, not user-input based). Unit tests should catch any invalid decisions.